dev-docs/RFCs/v7.3/data-management-rfc.md

```diff
@@ -0,0 +1,96 @@
+
+The lack of a global mechanism for managing data resources also has performance implications in that buffers and textures can be allocated to store redundant data. In deck.gl this might occur due to multiple layers being created from the same data (where each layer would allocate its own gl resources) or multiple instances of the same layer each creating their own layer-specific geometry buffers or texture data. The deck.gl `TileLayer` uses a `TileCache` to avoid redundantly creating textures, and a similar mechanism could be applied more generally to avoid allocating textures and buffers if like resources with the same data have already been created anywhere in the application.
+
+
+## Overview
+
+The proposed data manager would allow the application to register arbitrary named data sets, optionally providing accessors to parse particular views of the data. It would provide API methods to create buffers or textures from the
+registered data, returning cached versions if they are available.
+
+
+## Implementation
+
+A `DataManager` class that supports the following methods:
+- `addData(id, data, accessors)`: add a data set to the manager, optionally with accessors to create particular views.
+- `getBuffer(dataId, viewId)`: create a buffer from a data set view. On first call, create the buffer and cache it, subsequently return cached version and increment usage count. If `viewId` is omitted, the entire data set is used.
+- `getTexture(dataId, viewId)`: create a texture from a data set view. On first call, create the texture and cache it, subsequently return cached version and increment usage count. If `viewId` is omitted, the entire data set is used.
+- `release(Buffer|Texture)`: indicate that a resource is no longer used. Decrements the usage count for the resource and deletes it if the usage count reaches 0.
+
+
+## Example
+
+```js
+const dm = new DataManager(gl);
+
+dm.addData('myTable', [
+  {
+    position: [0.4, 1.0],
+    color: [255, 0, 255]
+  },
+  {
+    position: [1.0, 2.0],
+    color: [0, 255, 255]
+  },
+  {
+    position: [-1.0, 3.0],
+    color: [255, 0, 0]
+  }
+], {
+  positions: (data) => {
+    const result = new Float32Array(data.length * 2);
+    for (let i = 0, len = data.length; i < len; ++i) {
+      result.set(data[i].position, i * 2);
+    }
+    return result;
+  },
+  colors: (data) => {
+    const result = new Uint8Array(data.length * 3);
+    for (let i = 0, len = data.length; i < len; ++i) {
+      result.set(data[i].color, i * 3);
+    }
+    return result;
+  }
+});
+
+dm.addData('myImages', [
+  image1,
+  image2
+], {
+  image1: (data) => data[0],
+  image2: (data) => data[1]
+});
+
+dm.addData('anotherImage', image3); // Accessors not provided because the only view is the entire data set
+
+const positionBuffer = dm.getBuffer('myTable', 'positions');
+const positionBuffer2 = dm.getBuffer('myTable', 'positions'); // Gets cached version of same buffer
```

Comment:
In deck.gl `Attribute` class we need the following operations:

- Allocate an empty buffer of a specific size
- Create a new buffer from TypedArray
- Resize an existing buffer (keep the data but allocate bigger size)
- Partially update an existing buffer
- Copy a chunk from one buffer to another

Reply:
These should all be doable.